File: src/main/java/com/github/frcsty/frozenactions/util/Extensions.java
package com.github.frcsty.frozenactions.util;

import com.github.frcsty.frozenactions.wrapper.ActionHandler;
import me.clip.placeholderapi.PlaceholderAPI;
import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.minimessage.MiniMessage;
import net.kyori.adventure.text.minimessage.tag.resolver.TagResolver;
import org.bukkit.Bukkit;
import org.bukkit.entity.Player;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public final class Extensions {

    public static @NotNull String colorize(final @NotNull String text) {
        return ActionHandler.getAdventureMessage().parse(text).toString();
    }

    public static void sendTranslatedMessage(final @NotNull Player player, final @NotNull String text) {
        final String message = getTranslatedMessage(text, player);

        player.sendMessage(ActionHandler.getAdventureMessage().parse(message).toString());
    }

    public static @NotNull String getTranslatedMessage(@NotNull String text, final @NotNull Player player) {
        final boolean daddy = Bukkit.getPluginManager().isPluginEnabled("PlaceholderAPI");

        if (daddy)
            text = PlaceholderAPI.setPlaceholders(player, text);

        return text;
    }

    public static @NotNull Component getMiniTranslatedMessage(@NotNull String text, final @NotNull Player player) {
        final boolean daddy = Bukkit.getPluginManager().isPluginEnabled("PlaceholderAPI");

        if (daddy)
            text = PlaceholderAPI.setPlaceholders(player, text);

        return MiniMessage.miniMessage().deserialize(text);
    }

    public static @NotNull Component getMiniTranslatedMessage(@NotNull String text, final @NotNull Player player, final @Nullable TagResolver tags) {
        final boolean daddy = Bukkit.getPluginManager().isPluginEnabled("PlaceholderAPI");

        if (daddy)
            text = PlaceholderAPI.setPlaceholders(player, text);

        return MiniMessage.miniMessage().deserialize(text, tags);
    }

}

File: src/main/kotlin/com/github/frcsty/frozenactions/util/Extensions.kt
package com.github.frcsty.frozenactions.util

import me.clip.placeholderapi.PlaceholderAPI
import net.md_5.bungee.api.ChatColor
import net.md_5.bungee.api.chat.TextComponent
import org.bukkit.Bukkit
import org.bukkit.entity.Player
import org.bukkit.plugin.Plugin
import java.util.regex.Pattern

private val HEX_PATTERN: Pattern = Pattern.compile("#<([A-Fa-f0-9]){6}>")

fun String.color(): String {
    var translation = this

    if (Reflection.is16) {
        var matcher = HEX_PATTERN.matcher(translation)

        while (matcher.find()) {
            var hexString = matcher.group()

            hexString = "#" + hexString.substring(2, hexString.length - 1)
            val hex: ChatColor = ChatColor.of(hexString)
            val before = translation.substring(0, matcher.start())
            val after = translation.substring(matcher.end())

            translation = before + hex + after
            matcher = HEX_PATTERN.matcher(translation)
        }
    }

    return ChatColor.translateAlternateColorCodes('&', translation)
}

fun Player.sendTranslatedMessage(msg: String) {
    val message = msg.getTranslatedMessage(this)

    if (Reflection.is16) {
        this.spigot().sendMessage(*TextComponent.fromLegacyText(message.color()))
        return
    }

    this.sendMessage(message.color())
}

fun String.getTranslatedMessage(player: Player): String {
    var message = this
    val daddy: Plugin? = Bukkit.getPluginManager().getPlugin("PlaceholderAPI")

    if (daddy != null && daddy.isEnabled) {
        message = PlaceholderAPI.setPlaceholders(player, message)
    }

    return message
}